llvm/lib/Target/AMDGPU/AMDGPUEarlyRegisterSpilling.h

```diff
@@ -0,0 +1,177 @@
+  SetVectorType
+  collectUsesDominatedBySpill(Register DefRegToSpill,
+                              MachineInstr *SpillInstruction,
+                              const SetVectorType &UnreachableUses);
+
+  /// Find the restore locations, emit the restore instructions and maintain
+  /// SSA when needed.
+  void emitRestores(Register DefRegToSpill, MachineInstr *CurMI,
+                    MachineInstr *SpillInstruction,
+                    const SetVectorType &UnreachableUses,
+                    const TargetRegisterClass *RC, int FI);
+
+  /// Main spill functions that emits the spill and restore code.
+  void spill(MachineInstr *CurMI, GCNDownwardRPTracker &RPTracker,
+             unsigned NumOfSpills);
+
+  /// Emit restore instructions where it is needed
+  MachineInstr *emitRestore(Register SpillReg, MachineInstr *UseMI, int FI);
+  /// Emit restore instruction at the end of a basic block
+  MachineInstr *emitRestore(Register SpillReg, MachineBasicBlock *InsertBB,
+                            int FI);
+
+  /// Emit restore instructions for each group that contains the uses that are
+  /// dominated by the head of the group.
+  void emitRestoreInstrsForDominatedUses(
+      Register DefRegToSpill, MachineInstr *SpillInstruction,
+      MachineInstr *CurMI, SetVectorType &DominatedUses,
+      SmallVector<MachineInstr *> &RestoreInstrs,
+      SmallVector<MachineInstr *> &RestoreUses, int FI);
+
+  /// Check if it is legal or profitable to emit a restore in the common
+  /// dominator.
+  bool shouldEmitRestoreInCommonDominator(
+      MachineBasicBlock *SpillBlock, MachineBasicBlock *CurMBB,
+      MachineBasicBlock *CommonDominatorToRestore);
+
+  /// Find the common dominator of the reachable uses and the block that we
+  /// intend to spill.
+  MachineBasicBlock *
+  findCommonDominatorToSpill(MachineBasicBlock *SpillBlock,
+                             Register DefRegToSpill,
+                             const SetVectorType &ReachableUses);
+
+  /// Collect Reachable and Unreachable uses.
+  std::pair<SetVectorType, SetVectorType>
+  collectReachableAndUnreachableUses(MachineBasicBlock *SpillBlock,
+                                     Register DefRegToSpill,
+                                     MachineInstr *CurMI);
+
+  /// Helper functions to update the live interval analysis which is used by
+  /// the Register Pressure Tracker.
+  void updateIndexes(MachineInstr *MI);
+  void updateLiveness(Register Reg);
+  void updateLiveness(MachineInstr *MI);
+
+  bool hasPHIUseInSameBB(Register Reg, MachineBasicBlock *MBB);
+
+  /// Calculate the initial maximum register pressure per basic block (before
+  /// any spilling) because it gives us the maximum number of VGPRs and SGPRs
+  GCNRegPressure getMaxPressure(const MachineFunction &MF);
+
+  bool isSpilledReg(Register Reg) {
+    auto It = SpilledRegs.find(Reg);
+    if (It != SpilledRegs.end())
+      return true;
+    return false;
```

Comment:
```suggestion
    return It != SpilledRegs.end(); 
```

Instead of the whole if (ditto below)

Reply:
Good catch! @jmmartinez did a similar comment and I chose his proposal.